app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletFragment.kt

```diff
@@ -0,0 +1,146 @@
+    }
+
+    override val presenter: WalletContract.Presenter by inject()
+
+    private val binding: FragmentWalletBinding by viewBinding()
+
+    private val receiveAnalytics: ReceiveAnalytics by inject()
+
+    private lateinit var layoutManager: LinearLayoutManager
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        binding.setupView()
+
+        lifecycle.addObserver(presenter)
+    }
+
+    override fun showAddressCopied(addressOrUsername: String) {
+        requireContext().copyToClipBoard(addressOrUsername)
+        showUiKitSnackBar(
+            message = getString(R.string.home_address_snackbar_text),
+            actionButtonResId = R.string.common_ok,
+            actionBlock = Snackbar::dismiss
+        )
+    }
+
+    override fun showActionButtons(buttons: List<ActionButton>) {
+        binding.viewActionButtons.showActionButtons(buttons)
+    }
+
+    private fun FragmentWalletBinding.setupView() {
+        layoutManager = HomeScreenLayoutManager(requireContext())
+        layoutToolbar.setupToolbar()
+
+        homeRecyclerView.doOnAttach {
+            homeRecyclerView.layoutManager = layoutManager
+        }
+        homeRecyclerView.doOnDetach {
+            homeRecyclerView.layoutManager = null
+        }
+        swipeRefreshLayout.setOnRefreshListener(presenter::refreshTokens)
+        viewActionButtons.onButtonClicked = ::onActionButtonClicked
+
+        if (BuildConfig.DEBUG) {
+            with(layoutToolbar) {
+                viewDebugShadow.isVisible = true
+                imageViewDebug.isVisible = true
+                imageViewDebug.setOnClickListener {
+                    replaceFragment(DebugSettingsFragment.create())
+                }
+            }
+        }
+    }
+
+    private fun LayoutHomeToolbarBinding.setupToolbar() {
+        textViewAddress.setOnClickListener {
+            receiveAnalytics.logAddressOnMainClicked()
+            presenter.onAddressClicked()
+        }
+        imageViewProfile.setOnClickListener { presenter.onProfileClick() }
+        imageViewQr.setOnClickListener { replaceFragment(ReceiveSolanaFragment.create(token = null)) }
+    }
+
+    private fun onActionButtonClicked(clickedButton: ActionButton) {
+        when (clickedButton) {
+            ActionButton.SEND_BUTTON -> {
+                presenter.onSendClicked(clickSource = SearchOpenedFromScreen.MAIN)
+            }
+            ActionButton.SELL_BUTTON -> {
+                presenter.onSellClicked()
+            }
+            ActionButton.SWAP_BUTTON -> {
+                presenter.onSwapClicked()
+            }
+            ActionButton.TOP_UP_BUTTON -> {
+                presenter.onTopupClicked()
+            }
+            else -> {
+                // unsupported on this screen
+            }
+        }
+    }
+
+    override fun showUserAddress(ellipsizedAddress: String) {
```

Comment:
what if you create a ViewState here and just render(viewState) and then call all this methods?

Reply:
nihua ne ponyal